### PR TITLE
Tf 1.0.0

### PR DESCRIPTION
**Pull request by lissyx**

### CODE REVIEW

**Comment by kdavis-mozilla** [comment]
What about this for a plan....

Wait until the test of reuben's PR completes, then, before merging reuben's PR, test this PR, before we merge it, on the server with a 1.0.0 version of TensorFlow in your virtual env.

Maybe we'll get lucky and 1.0.0 will fix our problem and we won't have to revert the refactor.

Thoughts?

**Comment by lissyx** [comment]
Feels like a good plan to me ! 👍

File: DeepSpeech.py [hunk 1]
```diff
@@ -454,13 +455,12 @@ def get_tower_results(batch_set, optimizer=None):
...
-                total_loss, avg_loss, distance, accuracy, decoded, labels = \
-                    calculate_accuracy_and_loss(batch_set, no_dropout if optimizer is None else dropout_rates)
-
                 # Allow for variables to be re-used by the next tower
-                tf.get_variable_scope().reuse_variables()
+                with tf.variable_scope(tf.get_variable_scope(), reuse=True if i > 0 else None):
+                    # Calculate the avg_loss and accuracy and retrieve the decoded
+                    # batch along with the original batch's labels (Y) of this tower
+                    total_loss, avg_loss, distance, accuracy, decoded, labels = \
+                        calculate_accuracy_and_loss(batch_set, no_dropout if optimizer is None else dropout_rates)
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Maybe it's just me, but I don't understand this change at all. It seems un-related to the TF 0.12 to TF 1.0.0 migration.

In particular, [cifar10_multi_gpu_train.py](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py), the canonical example for multi-gpu code, handles multi-gpu code in the manner our old code did[[1](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py#L164)]. So I don't understand why this change is required.

File: DeepSpeech.py [hunk 2]
```diff
@@ -788,12 +788,12 @@ def create_execution_context(set_name):
             apply_gradient_op = apply_gradients(optimizer, avg_tower_gradients)
 
         # Create a saver to checkpoint the model
-        saver = tf.train.Saver(tf.all_variables())
+        saver = tf.train.Saver(tf.global_variables(), write_version=tf.train.SaverDef.V2)
```
Comment (lissyx) [inline DeepSpeech.py]: This is cherry-picked from your PR #106 :)

File: DeepSpeech.py (hunk 1 above)
Comment (lissyx) [inline DeepSpeech.py]: Without that change, code would not run and fail because of VarScope.

Comment (lissyx) [inline DeepSpeech.py]: "ValueError: Variable b1/Adam/ does not exist, or was not created with tf.get_variable(). Did you mean to set reuse=None in VarScope?"

File: DeepSpeech.py (hunk 2 above)
Comment (kdavis-mozilla) [inline DeepSpeech.py]: But that doesn't mean it's right :-)

File: DeepSpeech.py (hunk 1 above)
Comment (lissyx) [inline DeepSpeech.py]: BTW they did mostly the same, at https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_multi_gpu_train.py#L165: `with tf.variable_scope(tf.get_variable_scope()):`

I applied the same and updated the patch, that seems to run on my laptop with LDC93S1, fixing the VarScope that I was referring to above.

**Pull request merged**